DDD-7.md

```diff
@@ -0,0 +1,153 @@
+Number of worker threads will be specified either by the user via configuration or the default value will be used.
+The default value will be the number of the cores of the underlying machine.
+There will be only one thread pool, the one for processing CDC event pipelines.
+`SourceTasks`s, which will be also run in separate threads, start and stop with the engine and therefore don't need any thread pooling.
+
+### Running source tasks in parallel
+
+Kafka Connect `SourceTask` life cycle will be separated into a self-contained task.
+These tasks will be executed in dedicated threads created by [`Executors.newSingleThreadExecutor()`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/Executors.html#newSingleThreadExecutor()).
+In case of `RetriableException` the task will be restarted.
+In other cases, all other tasks will be stopped gracefully and the engine will fail with the task exception.
+
+### Processing CDC events concurrently
+
+Current `DebeziumEngine` API partially delegates event processing to the [`ChangeConsumer`](https://github.com/debezium/debezium/blob/v2.4.0.Final/debezium-api/src/main/java/io/debezium/engine/DebeziumEngine.java#L159) implementation, which is provided by the user.
+More specifically, events are passed to the [`ChangeConsumer#handleBatch()`](https://github.com/debezium/debezium/blob/v2.4.0.Final/debezium-api/src/main/java/io/debezium/engine/DebeziumEngine.java#L167) in a batch manner, as a `List`.
+This makes sense as in many cases events are submitted to another system and this is usually more efficient to do in a batch.
+However, this prevents us from making a complete event processing pipeline, which would be run in a dedicated thread.
+
+Before passing a batch of events to the `ChangeConsumer`, user defined single message transformations are applied to the records.
+In the case of [`ConvertingEngineBuilder`](https://github.com/debezium/debezium/blob/v2.4.0.Final/debezium-embedded/src/main/java/io/debezium/embedded/ConvertingEngineBuilder.java#L53), events are moreover serialized to `SourceRecords` before passing to the `ChangeConsumer`.
+Both these tasks will be run in parallel in `ThreadPoolExecutor` mentioned above.
+Before passing the batch to the `ChangeConsumer`, tasks for all records in the batch will be awaited to finish.
+
+The implementation should also try to optimize record serialization/deserialization in case of the `ConvertingEngineBuilder`.
+Currently, the records are serialized to `SourceRecords` before passing to the `ChangeConsumer` and then deserialized back when `ChangeConsumer` calls `RecordCommitter`.
+TBD task is to find a way how to avoid record deserialization during commit without breaking existing API.
+
+#### Disabling total ordering of the CDC events
```

Comment:
There should be a note about offsets handling. We need to make sure with both parallel processing and even more with unordered delivery that only most earlier offsets are comitted. There should be two shutdown modes - one that will process remaining messages up to the latest delivered and another one immediate whose result will be increased number of duplicates upon connector restarts.

Reply:
Good point - actually this is IMHO the hardest part of the whole implementation and it's not mentioned here as I haven't any good answer for it yet (I should probably add TBD section for it)